jmbitcoin/jmbitcoin/secp256k1_main.py

```diff
@@ -1,104 +1,234 @@
 #!/usr/bin/python
-from __future__ import print_function
+from __future__ import (absolute_import, division,
+                        print_function, unicode_literals)
+from builtins import *
+from future.utils import bytes_to_native_str
 import binascii
 import hashlib
 import re
 import sys
 import base64
+import struct
 import coincurve as secp256k1
 
 #Required only for PoDLE calculation:
 N = 115792089237316195423570985008687907852837564279074904382605163141518161494337
 
 #Standard prefix for Bitcoin message signing.
-BITCOIN_MESSAGE_MAGIC = '\x18' + 'Bitcoin Signed Message:\n'
-
-if sys.version_info.major == 2:
-    string_types = (str, unicode)
-    string_or_bytes_types = string_types
-    int_types = (int, float, long)
-
-    # Base switching
-    code_strings = {
-        2: '01',
-        10: '0123456789',
-        16: '0123456789abcdef',
-        32: 'abcdefghijklmnopqrstuvwxyz234567',
-        58: '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz',
-        256: ''.join([chr(x) for x in range(256)])
-    }
-
-    def lpad(msg, symbol, length):
-        if len(msg) >= length:
-            return msg
-        return symbol * (length - len(msg)) + msg
-
-    def get_code_string(base):
-        if base in code_strings:
-            return code_strings[base]
+BITCOIN_MESSAGE_MAGIC = b'\x18' + b'Bitcoin Signed Message:\n'
+
+string_types = (str)
+string_or_bytes_types = (str, bytes)
+int_types = (int, float)
+
+# Base switching
+code_strings = {
+    2: '01',
+    10: '0123456789',
+    16: '0123456789abcdef',
+    32: 'abcdefghijklmnopqrstuvwxyz234567',
+    58: '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz',
+    256: ''.join([chr(x) for x in range(256)])
+}
+
+def lpad(msg, symbol, length):
+    if len(msg) >= length:
+        return msg
+    return symbol * (length - len(msg)) + msg
+
+def get_code_string(base):
+    if base in code_strings:
+        return code_strings[base]
+    else:
+        raise ValueError("Invalid base!")
+
+def changebase(string, frm, to, minlen=0):
+    if frm == to:
+        return lpad(string, get_code_string(frm)[0], minlen)
+    return encode(decode(string, frm), to, minlen)
+
+def bin_to_b58check(inp, magicbyte=b'\x00'):
+    if not isinstance(magicbyte, int):
+        magicbyte = struct.unpack('B', magicbyte)[0]
+    if magicbyte == 0:
+        inp_fmtd = struct.pack('B', magicbyte) + inp
+    while magicbyte > 0:
+        inp_fmtd = struct.pack('B', magicbyte % 256) + inp
+        magicbyte //= 256
+    checksum = bin_dbl_sha256(inp_fmtd)[:4]
+    return b58encode(inp_fmtd + checksum)
```

Comment:
Not sure this function is behaving correctly.

Reply:
Behaviour appears correct except I didn't know that it's OK to have the magic/version byte be more than one byte? Or, more specifically, this function allows the magicbyte variable to be > 1 byte or 256 only if it's passed in as an integer. But I don't see incorrect behavior for the  1 byte case, as we use.